apps/dashboard/src/context/environment/environment-provider.tsx

```diff
@@ -35,18 +37,49 @@ function selectEnvironment(environments?: IEnvironment[], selectedEnvironmentId?
 
 export function EnvironmentProvider({ children }: { children: React.ReactNode }) {
   const { currentOrganization } = useAuth();
+  const navigate = useNavigate();
+  const { pathname } = useLocation();
   const [currentEnvironment, setCurrentEnvironment] = useState<IEnvironment | null>(null);
 
+  const switchEnvironmentInternal = useCallback(
+    (allEnvironments?: IEnvironment[], newEnvironmentId?: string) => {
+      const selectedEnvironment = selectEnvironment(allEnvironments, newEnvironmentId);
+      setCurrentEnvironment(selectedEnvironment);
+
+      if (
+        (pathname === ROUTES.ROOT || pathname === ROUTES.ENV || pathname === `${ROUTES.ENV}/`) &&
+        selectedEnvironment
+      ) {
+        navigate(buildRoute(ROUTES.WORKFLOWS, { environmentId: selectedEnvironment._id }));
+        return;
+      } else if (pathname.includes(ROUTES.ENV) && selectedEnvironment) {
+        const newPath = pathname.replace(/\/env\/[^/]+(\/|$)/, `${ROUTES.ENV}/${selectedEnvironment._id}$1`);
+        navigate(newPath);
+        return;
+      }
+
+      navigate(pathname, { replace: true });
```

Comment:
the navigation logic based on the current pathname handles the cases when the route is:
- `/`
- `/env`
- `/env/`
- `/env/:environmentId`
- `/env/:environmentId/`
- `/env/:environmentId/workflows`
- `/env/:environmentId/anything`

For the cases when the `environmentId` is present in the URL, it will replace it with the new environment id